src/fi_nomad/kernels/momentum_three_block_model_free.py

```diff
@@ -0,0 +1,144 @@
+        it with a momentum term (Block 1). It then updates candidate factor W (Block 2)
+        and H (Block 3). Finally the low-rank approximation (L) is calculated.
+        Extrapolation of L is done at the very beginning, starting with the second
+        iteration (instead of at the end of each step) to ensure L has rank r.
+
+        Returns:
+            None
+        """
+        if self.elapsed_iterations > 0:
+            # Seraghiti et. al. (2023) apply momentum at the end of each step
+            # but the last one, as this would alter the rank.  We omit an
+            # additional parameter "max iterations", by moving the momentum step
+            # on L to the beginning.
+            self.low_rank_candidate_L = apply_momentum(
+                self.low_rank_candidate_L,
+                self.previous_low_rank_candidate_L,
+                self.momentum_beta,
+            )
+
+        # BLOCK 1: construct utility Z and apply momentum term
+        utility_matrix_Z = construct_utility(
+            self.low_rank_candidate_L, self.sparse_matrix_X
+        )
+        self.utility_matrix_Z = apply_momentum(
+            utility_matrix_Z, self.utility_matrix_Z, self.momentum_beta
+        )
+
+        # BLOCK 2: update candidate factor W
+        self.candidate_factor_W = update_W(
+            self.candidate_factor_H, self.utility_matrix_Z
+        )
+
+        # BLOCK 3: update candidate factor H
+        self.candidate_factor_H = update_H(
+            self.candidate_factor_W, self.utility_matrix_Z
+        )
+
+        self.previous_low_rank_candidate_L = self.low_rank_candidate_L.copy()
+        self.low_rank_candidate_L = self.candidate_factor_W @ self.candidate_factor_H
+
+        if self.tolerance is not None:
+            self.loss = compute_loss(
+                self.utility_matrix_Z, self.low_rank_candidate_L, LossType.FROBENIUS
+            )
+
+    def running_report(self) -> str:
+        """Reports the current iteration number and loss. Only operative if a
+        tolerance was set.
+
+        Returns:
+            Description of current iteration and loss.
+        """
+        txt = (
+            ""
+            if self.tolerance is None
+            else f"iteration: {self.elapsed_iterations} loss: {self.loss}"
+        )
+        return txt
+
+    def report(self) -> KernelReturnType:
+        """Returns final low-rank approximation and descriptive string.
+        The description indicates the total number of iterations completed,
+        and the final loss (if a tolerance was set).
+
+        Returns:
+            Object containing results and summary
+        """
+        floss = str(self.loss) if self.loss != float("inf") else "Not Tracked"
+        text = f"{self.elapsed_iterations} total, final loss {floss}"
+        data = Momentum3BlockModelFreeKernelReturnType(
+            (self.candidate_factor_W, self.candidate_factor_H)
+        )
+        return KernelReturnType(text, data)
```

Comment:
I saw your note that this is repeated and could be moved to the base class. That's a good idea. The only tricky part will be adding the `data` part, since it's not easy to do one-line differences in the middle of functions. But we can either make it work with inheritance or restructure how the report call gets made.

Let's make that cleanup a separate issue--I don't want this PR to be held up just for solving that.